addons/sale_margin/models/sale_order.py

```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+import odoo.addons.decimal_precision as dp
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    margin = fields.Float(compute='_product_margin', digits=dp.get_precision('Product Price'), store=True)
+    purchase_price = fields.Float(string='Cost Price', digits=dp.get_precision('Product Price'))
+
+    @api.onchange('product_id', 'product_uom_qty')
+    def product_id_change_margin(self):
+        if self.order_id.pricelist_id:
+            frm_cur = self.env.user.company_id.currency_id
+            to_cur = self.order_id.pricelist_id.currency_id
+            purchase_price = self.product_id.standard_price
+            if self.product_uom != self.product_id.uom_id:
+                purchase_price = self.env['product.uom']._compute_price(self.product_id.uom_id.id, purchase_price, to_uom_id=self.product_uom.id)
+            price = frm_cur.with_context(date=self.order_id.date_order).compute(purchase_price, to_cur, round=False)
+            self.purchase_price = price
+
+    @api.depends('purchase_price', 'product_uom_qty', 'price_unit')
+    def _product_margin(self):
+        for line in self:
+            currency = line.order_id.pricelist_id.currency_id
+            line.margin = currency.round(line.price_subtotal - ((line.purchase_price or line.product_id.standard_price) * line.product_uom_qty))
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    margin = fields.Monetary(compute='_product_margin', help="It gives profitability by calculating the difference between the Unit Price and the cost price.", currency_field='currency_id', digits=dp.get_precision('Product Price'), store=True)
+
+    @api.depends('order_line.margin')
+    def _product_margin(self):
+        for order in self:
+            order.margin = 0.0
+            for line in order.order_line:
+                if line.filtered(lambda r: r.state == 'cancel'):
+                    continue
```

Comment:
Use: for line in order.order_line.filtered(lambda r: r.state != 'cancel') instead of above statement .